aips/on-chain-randomness-impl.md

```diff
@@ -0,0 +1,419 @@
+*On-chain configuration changes.* For periodic reconfigurations that happen every two hours, the validators need to finish the DKG before the next epoch starts. For governance proposals that require reconfiguration, the changes by the governance proposal will be buffered on-chain, and applied after the DKG finishes and validators enter the new epoch. Any future new on-chain configs need to follow the same procedure.
+*Validator set changes.* In the current design, during the DKG phase any changes to the `ValidatorSet` will be rejected. Based on the latest mainnet simulation, it means for up to 30 secs in every 2 hours, transactions for validator set changes will fail. 
+
+More details can be found in the Specification section.
+
+**Constraints on validator set size**. If the validator set size is too large, in the worst case when the stake distribution is adversarial, the rounding algorithm will output DKG parameters with large total weight, causing the DKG transcript to exceed some transaction limits. When this happens, the DKG may lose liveness. 
+
+Here are some related transaction limits. As a prevention, the DKG transaction size will be monitored closely and the corresponding alerts will be set properly. When the DKG transaction size is close to the limit, we may have to relax the transaction limits below. 
+
+- `max_bytes_per_write_op`: limit of the size of a single state item, 1MB currently.
+- `ValidatorTxnConfig::V1::per_block_limit_total_bytes`: limit of the total validator transaction size in one block, 2MB currently.
+
+**Minimum gas deposit**. To prevent [undergasing attack](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-41.md#undergasing-attacks), the current implementation requires randomness transactions to have a minimum balance of 2 APT to deposit enough gas in the transaction execution prologue, to ensure that execution will not be aborted after the randomness call. More specifically, 
+- New steps in a user transaction execution.
+  - Before prologue, determine if the txn requires deposit, if so how much.
+  - At the end of prologue, if deposit is required, burn the calculated amount from the gas payer's balance.
+  - At the beginning of epilogue, if deposit was required, mint and refund the same amount back to the gas payer.
+- For randomness transactions, define the required deposit as `min(A, B)`, where:
+  -  `A = (max_execution_gas + max_io_gas) * price_per_gas_unit + max_storage_fee`, and
+  -  `B = maximum_number_of_gas_units * price_per_gas_unit`.
+
+Some reference numbers on mainnet.
+```
+max_execution_gas=920
+max_io_gas=1_000
+typical_octa_per_gas=100
+max_storage_fee=200_000_000
+max_number_of_gas_units=2_000_000
+```
+Therefore, for a txn that directly or indirectly call randomness API, the required deposit is 2_000_000 octas, i.e., 2APT.
+
+TODO: add / move details about undergasing attack prevent once implementation is finalized
```

Comment:
cc @zjma

Reply:
based on the discussion today, we probably proceed with the current implementation.